flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/JoinToMultiJoinRule.java

```diff
@@ -0,0 +1,683 @@
+ * reordering and will be turned into a {@link
+ * org.apache.flink.table.planner.plan.nodes.physical.stream.StreamPhysicalMultiJoin}.
+ *
+ * <p>Join conditions are also pulled up from the inputs into the topmost {@link MultiJoin}.
+ *
+ * <p>Join information is also stored in the {@link MultiJoin}. Join conditions are stored in arrays
+ * in the {@link MultiJoin}. This outer join information is associated with the null generating
+ * input in the outer join. So, in the case of a left outer join between A and B, the information is
+ * associated with B, not A.
+ *
+ * <p>Here are examples of the {@link MultiJoin}s constructed after this rule has been applied on
+ * following join trees.
+ *
+ * <ul>
+ *   <li>A JOIN B &rarr; MJ(A, B)
+ *   <li>A JOIN B JOIN C &rarr; MJ(A, B, C)
+ *   <li>A LEFT JOIN B &rarr; MJ(A, B)
+ *   <li>A RIGHT JOIN B &rarr; MJ(A, B)
+ *   <li>A FULL JOIN B &rarr; MJ[full](A, B)
+ *   <li>A LEFT JOIN (B JOIN C) &rarr; MJ(A, B, C)
+ *   <li>(A JOIN B) LEFT JOIN C &rarr; MJ(A, B, C)
+ *   <li>(A LEFT JOIN B) JOIN C &rarr; MJ(A, B, C)
+ *   <li>(A LEFT JOIN B) LEFT JOIN C &rarr; MJ(A, B, C)
+ *   <li>(A RIGHT JOIN B) RIGHT JOIN C &rarr; MJ(MJ(A, B), C)
+ *   <li>(A LEFT JOIN B) RIGHT JOIN C &rarr; MJ(MJ(A, B), C)
+ *   <li>(A RIGHT JOIN B) LEFT JOIN C &rarr; MJ(MJ(A, B), C)
+ *   <li>A LEFT JOIN (B FULL JOIN C) &rarr; MJ(A, MJ[full](B, C))
+ *   <li>(A LEFT JOIN B) FULL JOIN (C RIGHT JOIN D) &rarr; MJ[full](MJ(A, B), MJ(C, D))
+ *   <li>SEMI JOIN and ANTI JOIN not support now.
+ * </ul>
+ *
+ * <p>The constructor is parameterized to allow any sub-class of {@link Join}, not just {@link
+ * LogicalJoin}.
+ *
+ * @see FilterMultiJoinMergeRule
+ * @see ProjectMultiJoinMergeRule
+ * @see CoreRules#JOIN_TO_MULTI_JOIN
+ */
+@Value.Enclosing
+public class JoinToMultiJoinRule extends RelRule<JoinToMultiJoinRule.Config>
+        implements TransformationRule {
+
+    public static final JoinToMultiJoinRule INSTANCE = JoinToMultiJoinRule.Config.DEFAULT.toRule();
+
+    /** Creates a JoinToMultiJoinRule. */
+    public JoinToMultiJoinRule(Config config) {
+        super(config);
+    }
+
+    @Deprecated // to be removed before 2.0
+    public JoinToMultiJoinRule(Class<? extends Join> clazz) {
+        this(Config.DEFAULT.withOperandFor(clazz));
+    }
+
+    @Deprecated // to be removed before 2.0
+    public JoinToMultiJoinRule(
+            Class<? extends Join> joinClass, RelBuilderFactory relBuilderFactory) {
+        this(
+                Config.DEFAULT
+                        .withRelBuilderFactory(relBuilderFactory)
+                        .as(Config.class)
+                        .withOperandFor(joinClass));
+    }
+
+    // ~ Methods ----------------------------------------------------------------
+
+    @Override
+    public boolean matches(RelOptRuleCall call) {
+        final Join origJoin = call.rel(0);
+        if (origJoin.getJoinType() != JoinRelType.INNER
```

Comment:
I remember we had some issues with the multi way join rules and `FOR SYSTEM TIME AS OF`, is it guaranteed that lookups joins are not merged into multijoin?

Reply:
Potentially also interval joins should not be pulled into the multijoin.

Reply:
Yes, we avoid matching temporal joins. I basically had done the same we do for binary joins, see https://github.com/apache/flink/blob/d87ff8c3ff0508e9e3b82630d59f07b27c4ecbcd/flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/physical/stream/StreamPhysicalMultiJoinRule.java#L71. 

I took a look again and I agree this doesn't look complete. I've added the necessary checks in the logical rule + tests for it, see https://github.com/apache/flink/pull/26687/commits/e0ccc9c1d62fcb8a129a7eb5354ea6f2f40fbbc2

Reply:
If this look good to you I could add this + tests for the reorder rule in another ticket